examples/StreamLayerReadExample.cpp

```diff
@@ -0,0 +1,196 @@
+          }
+        } else {
+          OLP_SDK_LOG_INFO_F(kLogTag, "Message data: size - %lu",
+                             msg.GetData()->size());
+        }
+      }
+    } else {
+      OLP_SDK_LOG_INFO(kLogTag, "No new messages is received");
+    }
+  } while (messages_size > 0);
+  return 0;
+}
+
+int DeleteSubscription(olp::dataservice::read::StreamLayerClient& client) {
+  auto unsubscribe_future = client.Unsubscribe();
+  auto unsubscribe_response = unsubscribe_future.GetFuture().get();
+  if (!unsubscribe_response.IsSuccessful()) {
+    OLP_SDK_LOG_ERROR_F(kLogTag,
+                        "Failed to unsubscribe - HTTP Status: %d Message: %s",
+                        unsubscribe_response.GetError().GetHttpStatusCode(),
+                        unsubscribe_response.GetError().GetMessage().c_str());
+    return -1;
+  }
+  return 0;
+}
+}  // namespace
+
+int RunStreamLayerExampleRead(
+    const AccessKey& access_key, const std::string& catalog,
+    const std::string& layer_id,
+    const boost::optional<
+        olp::dataservice::read::SubscribeRequest::SubscriptionMode>&
+        subscription_mode) {
+  // Create a task scheduler instance
+  std::shared_ptr<olp::thread::TaskScheduler> task_scheduler =
+      olp::client::OlpClientSettingsFactory::CreateDefaultTaskScheduler();
+  // Create a network client
+  std::shared_ptr<olp::http::Network> http_client = olp::client::
+      OlpClientSettingsFactory::CreateDefaultNetworkRequestHandler();
+
+  // Initialize authentication settings
+  olp::authentication::Settings settings({access_key.id, access_key.secret});
+  settings.task_scheduler = std::move(task_scheduler);
+  settings.network_request_handler = http_client;
+  // Setup AuthenticationSettings with a default token provider that will
+  // retrieve an OAuth 2.0 token from OLP.
+  olp::client::AuthenticationSettings auth_settings;
+  auth_settings.provider =
+      olp::authentication::TokenProviderDefault(std::move(settings));
+
+  // Setup OlpClientSettings and provide it to the StreamLayerClient.
+  olp::client::OlpClientSettings client_settings;
+  client_settings.authentication_settings = auth_settings;
+  client_settings.network_request_handler = std::move(http_client);
+
+  // Create subscription, used kSerial or kParallel subscription mode
+  olp::dataservice::read::SubscribeRequest subscribe_request;
+  if (subscription_mode) {
+    subscribe_request.WithSubscriptionMode(subscription_mode.get());
+  }
+
+  auto read_from_stream_layer = [&]() -> int {
+    // Create stream layer client with settings and catalog, layer specified
+    olp::dataservice::read::StreamLayerClient client(olp::client::HRN{catalog},
+                                                     layer_id, client_settings);
+    int res = 0;
+    if (CreateSubscription(client, subscribe_request) != 0 ||
+        RunPoll(client) != 0) {
+      res = -1;
+    }
+    if (DeleteSubscription(client) != 0) {
+      res = -1;
+    }
+    return res;
+  };
```

Comment:
This looks very cryptic. Is there no better way to do it? Looking at this code I really don't understand how to use the stream layer client to poll data. Please try to keep the overhead, e.g. logs and return values as slim as possible. You can simply return bool from functions that only use 0 or -1.

Here is an example from the Kafka CPP client -> https://github.com/mfontanini/cppkafka/blob/master/examples/consumer_example.cpp